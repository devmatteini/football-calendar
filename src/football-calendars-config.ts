import * as Effect from "effect/Effect"
import * as S from "effect/Schema"
import * as FileSystem from "@effect/platform/FileSystem"
import * as F from "effect/Function"
import * as Config from "effect/Config"
import * as os from "node:os"
import * as path from "node:path"
import { parseJsonFile } from "./common/file"

export const Team = S.TaggedStruct("Team", {
    teamId: S.Int,
    name: S.String,
})
export type Team = typeof Team.Type

export const LeagueRound = S.Literal("All", "KnockoutStage")
export type LeagueRound = typeof LeagueRound.Type

export const League = S.TaggedStruct("League", {
    leagueId: S.Int,
    name: S.String,
    season: S.Int.pipe(S.positive()),
    round: S.optionalWith(LeagueRound, { default: () => "All" }),
})
export type League = typeof League.Type

export const FootballCalendar = S.Union(Team, League)
export type FootballCalendar = typeof FootballCalendar.Type

export const FootballCalendars = S.NonEmptyArray(FootballCalendar)
export type FootballCalendars = typeof FootballCalendars.Type

export const loadFootballCalendarConfig = Effect.gen(function* (_) {
    const fs = yield* _(FileSystem.FileSystem)

    const configFile = yield* _(configFilePath)
    const configExists = yield* _(fs.exists(configFile))
    if (!configExists) return yield* _(Effect.fail(`Configuration file doesn't exists (${configFile})`))

    return yield* _(parseJsonFile(configFile, FootballCalendars))
}).pipe(Effect.orDie)

const configJson = ["football-calendar", "config.json"] as const

const configFilePath = F.pipe(
    Config.string("FOOTBALL_CALENDAR_CONFIG"),
    Config.orElse(() =>
        Config.map(Config.string("XDG_CONFIG_HOME"), (configDir) => path.join(configDir, ...configJson)),
    ),
    Config.withDefault(path.join(os.homedir(), ".config", ...configJson)),
)
