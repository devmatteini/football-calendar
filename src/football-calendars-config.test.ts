import { expect, test } from "vitest"
import { FootballCalendarsJsonSchema } from "./football-calendars-config"

test("FootballCalendars JSON Schema - if this test fails the public API contract may be broken", () => {
    expect(FootballCalendarsJsonSchema).toMatchSnapshot()
})
