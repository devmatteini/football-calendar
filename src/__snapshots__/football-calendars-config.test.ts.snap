// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FootballCalendars JSON Schema - if this test fails the public API contract may be broken 1`] = `
{
  "$defs": {
    "Int": {
      "description": "an integer",
      "title": "int",
      "type": "integer",
    },
  },
  "$schema": "http://json-schema.org/draft-07/schema#",
  "items": {
    "anyOf": [
      {
        "additionalProperties": false,
        "properties": {
          "_tag": {
            "enum": [
              "Team",
            ],
            "type": "string",
          },
          "name": {
            "type": "string",
          },
          "teamId": {
            "$ref": "#/$defs/Int",
          },
        },
        "required": [
          "_tag",
          "teamId",
          "name",
        ],
        "type": "object",
      },
      {
        "additionalProperties": false,
        "properties": {
          "_tag": {
            "enum": [
              "League",
            ],
            "type": "string",
          },
          "leagueId": {
            "$ref": "#/$defs/Int",
          },
          "name": {
            "type": "string",
          },
          "round": {
            "enum": [
              "All",
              "KnockoutStage",
              "Final",
            ],
            "type": "string",
          },
          "season": {
            "$ref": "#/$defs/Int",
            "description": "a positive number",
            "exclusiveMinimum": 0,
            "title": "positive",
          },
        },
        "required": [
          "_tag",
          "leagueId",
          "name",
          "season",
        ],
        "type": "object",
      },
    ],
  },
  "minItems": 1,
  "type": "array",
}
`;
